### PR TITLE
E2E testing setup

### PR DESCRIPTION
**Pull request by gr812b**
Adding E2E Tests! Now we have our automated QA man to test it all. How lovely

**In this PR:**
- [X] Readme updates!
- [X] Testing upload form
- [x] Testing basic graph making
- [ ] Testing a few advanced graph features

This will not be exhaustive, as much of the front end error handling and UI will be rebuilt.

### CODE REVIEW

File: e2e/testers/graph.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+export class GraphTester extends Tester {
+
+  constructor(page: Page) {
+    super(page);
+  }
+
+  async testCreateGraph() {
+    await this.openGraphForm();
+    await this.pressNextButton();
+    await this.selectFile('182848/BATT VOLT');
```
Comment (gr812b) [inline e2e/testers/graph.ts:13]: Should be a variable

File: e2e/testers/upload.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      throw new Error(`Test file not found at ${this.binFilePath}`);
+    }
+  }
+
+  async testUploadBinFile() {
+    await this.openUploadForm();
+    await this.uploadFile(this.binFilePath);
+    await this.submitFile();
+    await this.verifyFilesUploaded();
+    await this.awaitFileParsed('182848/BATT VOLT');
```
Comment (gr812b) [inline e2e/testers/upload.ts:27]: Should be a variable

File: e2e/tests/create/no-analyzer.test.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { test, expect } from '@playwright/test';
+import { UploadTester, GraphTester } from '../../testers';
+
+test.describe('Create graphs', () => {
+
+  test.beforeEach(async ({ page }) => {
+    const uploadTester = new UploadTester(page);
+    await page.goto('http://localhost:5173/');
```
Comment (gr812b) [inline e2e/tests/create/no-analyzer.test.ts:8]: Make this a variable across files

**Review by gr812b — COMMENTED** [review]
Maybe add linter options here?

File: e2e/playwright.config.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    //   use: { ...devices['Desktop Chrome'], channel: 'chrome' },
+    // },
+  ],
+
+  /* Run your local dev server before starting the tests */
+  // webServer: {
+  //   command: 'npm run start',
+  //   url: 'http://127.0.0.1:3000',
+  //   reuseExistingServer: !process.env.CI,
+  // },
```
Comment (camdnnn) [inline e2e/playwright.config.ts:79]: Is there a reason for these commented out lines of code to be here

File: e2e/testers/graph.ts (hunk 1 above)
Comment (camdnnn) [reply e2e/testers/graph.ts:13]: Also for this test how is that bin file being fetched? Where is it being saved and loaded from?

File: e2e/test-files/182848.bin
[the remark this answers is not in the document]
Comment (camdnnn) [reply e2e/test-files/182848.bin:1]: there's probably not a better way I just think it would be better to maybe use a more purpose built bin file to account for all possibilities

**Review by camdnnn — APPROVED** [review]
Seems pretty good to me just think we should look into better ways of handling the files in the future

File: e2e/testers/graph.ts (hunk 1 above)
Comment (gr812b) [reply e2e/testers/graph.ts:13]: This function is for creating the graph exclusively, with the precondition that the file is uploaded. All of the functions in here are merely pieces to be put together to write full tests in the .test.ts files

File: e2e/playwright.config.ts (hunk 4 above)
Comment (gr812b) [reply e2e/playwright.config.ts:79]: This is just default with the setup for the playwright.config.ts, kept here because the option to automatically run the service before startup sounds maybe like something we might want in the future.

File: e2e/test-files/182848.bin
[the remark this answers is not in the document]
Comment (gr812b) [reply e2e/test-files/182848.bin:1]: I don't know what you mean by a purpose built file. The goal of E2E tests is to ensure the whole system / functions work, not really testing edge cases. Using a real .bin file will most likely be the best representation of this. In terms of referencing the file, I need to improve that for sure. I may define something in the testers folder to specify files and information about them.

File: e2e/test-files/endurance.bin
[the remark this answers is not in the document]
Comment (gr812b) [reply e2e/test-files/endurance.bin:1]: Good for bigger tests should it come to it. Maybe I'll write one that uses it.

File: data/2023-02-26/accelRun1Thick2ThinHeavy.csv
Comment (t-wing11) [inline data/2023-02-26/accelRun1Thick2ThinHeavy.csv:1]: Good idea to remove all this data, no need for all of it

File: e2e/test-files/endurance.bin
Comment (t-wing11) [inline e2e/test-files/endurance.bin:1]: nice using LFS is very cool

File: e2e/testers/upload.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    this.frontendFilePath = `${binFileName}/BATT VOLT`;
+
+    if (!fs.existsSync(this.binFilePath)) {
+      throw new Error(`Test file not found at ${this.binFilePath}`);
+    }
+  }
+
+  async testUploadBinFile() {
+    await this.openUploadForm();
+    await this.uploadFile(this.binFilePath);
```
Comment (t-wing11) [inline e2e/testers/upload.ts:26]: This is actually so sick, being able to nest tests like this

**Review by t-wing11 — COMMENTED** [review]
Some thoughts

- Would some sort of tracker to see what has been created and what needs to be created, perhaps a list or something that can check tests off when they are done until everything has been covered
- For these types of PR's going forward a video of the test would be nice so we can see that whatever was done actually works
- did you add both those tests to the actions? If so some documentation on how to do that would be good for future test creation
- This is good initial setup but we will need to monitor to make sure that no duplicate tester functions are made. No need to waste code if its already created.
- -400k lines gonna hurt ur analytics for this repo

**Review by t-wing11 — APPROVED** [review]

**Comment by gr812b** [comment]
> > Some thoughts
> > 
> > * Would some sort of tracker to see what has been created and what needs to be created, perhaps a list or something that can check tests off when they are done until everything has been covered
> > * For these types of PR's going forward a video of the test would be nice so we can see that whatever was done actually works
> > * did you add both those tests to the actions? If so some documentation on how to do that would be good for future test creation
> > * This is good initial setup but we will need to monitor to make sure that no duplicate tester functions are made. No need to waste code if its already created.
> > * -400k lines gonna hurt ur analytics for this repo
> 
> Actually @gr812b i can make the list of things that need to be tested. Maybe i'll make an excel sheet or smt

Oh god we're getting deep on nested replies. As far as the videos go, any E2E tests should probably just be pulled and tested using the UI. Mention of this in the PR would be great. Some PRs mght have many many tests so it might get out of hand adding videos.

Documentation will be added at some point to the wiki on all this.

This repos analytics are already fucked, just trying to recuperate what I can lol. 

In regards to the duplicate tester stuff, yea this stuff kind of needs an understanding of what features exist already and to use them rather than re-doing it. Should be alright but we'll see how it goes.
